File: BackEnd/Mmo/Mmo_Domain/ModelRequest/ProductRequest.cs
using System.ComponentModel.DataAnnotations;

namespace Mmo_Domain.ModelRequest;

public class ProductRequest
{
    [Required]
    public int? ShopId { get; set; }
    
    [Required]
    public int? CategoryId { get; set; }

    public int? SubcategoryId { get; set; }
    
    [Required]
    [MaxLength(100)]
    public string Name { get; set; } = null!;
    
    [Required]
    [MaxLength(500)]
    public string? Description { get; set; }

    public string? Details { get; set; }
    
    public List<ProductVariantRequest>? Variants { get; set; }
}

File: frontend/src/models/modelRequest/ProductRequest.ts
export interface ProductStorageRequest {
  result: string; // JSON string
}

export interface ProductVariantRequest {
  id?: number; // Variant ID for update operations
  name: string;
  price: number;
  stock?: number;
  storages?: ProductStorageRequest[];
  storageJson?: string; // Raw JSON string chứa mảng tài khoản
}

export interface ProductRequest {
  name: string;
  description: string;
  price: number;
  categoryId: number;
  subcategoryId?: number;
  shopId: number;
  stock: number;
  images: string[];
  isActive: boolean;
  details?: string;
  variants?: ProductVariantRequest[];
}

export interface ProductUpdateRequest extends ProductRequest {
  id: number;
}

export interface ProductListRequest {
  page?: number;
  limit?: number;
  search?: string;
  categoryId?: number;
  shopId?: number;
  isActive?: boolean;
  sortBy?: "name" | "price" | "createdAt" | "updatedAt";
  sortOrder?: "asc" | "desc";
}
